Improve the context menu actions when multiple TCOs are selected

IanCaio opened the pull request:
Right clicking a TCO on the Song Editor, we have access to actions like _Cut_, _Copy_, _Paste_, _Remove_ and _Mute_. When the user has multiple TCOs selected, it's expected that the actions will affect all TCOs, but currently they only affect the TCO that was right-clicked on. This draft starts working on a solution that will enable those actions to affect all selected TCOs.

In its current state, _Remove_ and _Mute_ are fully functional (they will affect the individual TCO if there are no TCOs selected, or all selected TCOs if there are). _Cut_ and _Copy_ are partially functional (they will create a Datafile with the TCO information and copy it to the Clipboard when multiple TCOs are selected, or do the _Cut_/_Copy_ the previous way if no TCOs are selected). _Paste_ is only functional for single TCOs (when multiple TCOs were previously copied/cut, it will print the copied data in the terminal using `qWarning()` and when a single TCO was copied/cut it will paste normally).

Reason I opened a draft is that it will be easier to allow others to test what is written so far with the LMMSBot builds and also because it will make it more visible so others can share their ideas on how to finish it. I'm still figuring out the _Paste_ operation.

Closes #1212

IanCaio committed:
Starts implementing the feature

The idea of this branch is to allow actions triggered through the context menu of a TCO on the song editor to affect all the selected TCOs. With this commit, only the "Mute/unmute" action affects all selected TCOs, while the others retain their old behavior (only affect the TCO that owns the context menu).

For that, a method was created that processes all actions (the triggered action is parsed as a parameter to the method). In the case of the "Mute" action, it checks if the song editor has selected TCOs, and if it does it mutes/unmutes all of them.

IanCaio committed:
Allows selected TCOs to be removed too

Now the "Remove" action from the context menu will remove all selected TCOs if there are any.

IanCaio committed:
Starts implementing selected TCO cut and copy

Now, when multiple TCOs are selected, the context menu actions Cut and Copy will write a DataFile to the clipboard containing the TCO information, so it can later be used to paste it.

The Paste action now checks if there's data in the QApplication clipboard. If there is, it will later paste the TCOs (for now it just prints the data with qWarning). If there's not, it uses the regular TCO paste method that uses the internal LMMS clipboard. Because it now have to decide between the QApplication clipboard and the LMMS internal clipboard, the Clipboard::copy() method now clears anything in the QApplication clipboard, making it empty so LMMS can know it should use the internal clipboard instead in that situation.

Adds safety checks for the selected TCO views.

IanCaio committed:
Merge branch 'master' into feature/selectedTCOActions

IanCaio committed:
Overloads TCW paste selection methods

This commit is a step towards implementing the paste feature of the TCO context menu. It overloads the TrackContentWidget::canPasteSelection and TrackContentWidget::pasteSelection methods so they can be called without having a QDropEvent (only the QMimeData with the copied TCOs information). The overloaded canPasteSelection(MidiTime, QMimeData, bool) method required a third argument which says whether pasting over the same bar should be allowed or not, because it shouldn't when the QDropEvent comes from the same application but should when it doesn't. That is defined in the canPasteSelection(MidiTime, QDropEvent) method.

Also, the pasteSelection(MidiTime, QDropEvent) isn't optimal, since it calls canPasteSelection twice (more details in the comments, but it's basically because the two canPasteSelection methods can return different values depending on the origin of QDropEvent). This could later be fixed by calling canPasteSelection before pasteSelection and removing it from inside the method altogether.

Next step is to add the "tco_" key to the mimeData on the Copy/Cut operations and implementing the paste operation using those methods.

IanCaio commented:
Methods used for the Drag&Drop paste were overloaded so they can be used with a QMimeData object instead of a QDropEvent one. That makes it possible for us to reuse those methods for our Context Menu paste actions. There are some details that are not optimal (`canPasteSelection` is called twice, once on every version of the method, if the `QDropEvent` `pasteSelection` method is called), but they work the way they are supposed to and could be fixed in a somewhat simple way.

IanCaio commented:
**Note to self:**

Apparently `src/tracks/SampleTrack.cpp` overrides the `contextMenuEvent` method. My changes end up not affecting the Sample tracks and even misbehaving. That file will also need to be changed appropriately.

IanCaio committed:
Adds the TCO type to the clipboard

Adds the key with the TCO type ("tco_" + type number + ":" + TCO Data XML) to the clipboard, so it can be later used by the canPasteSelection and pasteSelection methods.

IanCaio committed:
Apply changes to "src/tracks/SampleTrack.cpp"

Change the SampleTCOView::contextMenuEvent() method so it behaves the same way as the TrackContentObjectView::contextMenuEvent() method. For that, I had to change the ContextMenuAction enum and the contextMenuAction methods to be protected instead of private, so SampleTCOView can access it.

IanCaio committed:
Implement the paste action

Now that the canPasteSelection and pasteSelection methods were overloaded, it was possible to implement the paste action using the same logic as the Drag&Drop copy/paste.

Other small changes:
	- Removes the TCO views AFTER creating the TCO data file, instead of before (which probably resulted in an empty data file).
	- Uses the StringPairDrag::mimeType() instead of Clipboard::mimeType() since that's the one the methods from StringPairDrag.cpp recognizes.

IanCaio committed:
Removes QDebug header

Forgot to remove the QDebug header on the last commit.

IanCaio commented:
The basic functionality is implemented, all the Context Menu actions (Remove, Cut, Copy, Paste and Mute) will now affect all the TCOs selected. The Paste action follow the same rules as the Drag&Drop copy/paste:
- TCO being pasted to has to be of the same type of the TCO from where the context menu was opened
- All copied tracks need to land in a track of the same type
- You can't paste on the same bar where you copied

The current state of the PR is still a draft because there are probably things that should be changed to make the code better (one example is the issue with `canPasteSelection` being called twice on the `pasteSelection` that receives a `QDropEvent`). I also wish to create a context menu on the TrackContentWidget so the user can paste anywhere in the Track, instead of just over another Track. I can make the PR "Ready to review" though if the Devs would like to give input already on how things could be improved, just let me know!

IanCaio committed:
Creates a Context Menu on the TCW for "paste"

Now it's possible to paste a selection of copied TCOs anywhere in the TCW, as long as the rules to paste are met (same rules as Drag&Drop copy/paste). If the rules are not met the "Paste" menu action shows but is disabled.

IanCaio committed:
Small code refactoring

Saving a few lines of code.

IanCaio committed:
Avoids double call to canPasteSelection

This commit adds a third parameter to the pasteSelection overloaded method, which will define whether we whould skip the canPasteSelection check. The only situation where we will want to skip it is if we are calling pasteSelection from inside the other pasteSelection method, which will already have checked it. Because of that the default value is false.

Organizes comments as well.

IanCaio commented:
Marking the PR as Ready for review. All proposed functionality was implemented!

In my opinion, we should soon move away from using the `Clipboard` class and use the system's clipboard instead for all TCO copy/paste operations (and even others that currently use `Clipboard.h`, like `PianoRoll.cpp`). My PR uses the system's clipboard just like Drag&Drop copy/paste does. However, individual TCO copy/paste, which was already implemented, still uses the `Clipboard` class. It's possible to swift everything to the use the regular clipboard and I think we should do it soon, but IMHO in a different PR. Not only it's outside the scope of this one, but it would make it easier to review.

The functionality added should help a lot with people's workflow, since copying/pasting during production is a very constant thing and currently besides the Drag&Drop those operations are very limited.

russiankumar commented:
Alright, so I tested this, here's what I have to say:
- This works more or less as expected when using the menu
- While using shortucts (eg. for muting - Ctrl+middle click) it still affects only the targeted TCO and not all the selected TCOs
- Right-click menu for paste works a bit weird. Say I want to paste somewhere but I hit the wrong box by mistake. In that case I would move my mouse and right click on the correct box. However I have to right click twice for this as the first right click only disables the previous menu. _(On later testing, this is actually observed with every menu, so this is probably a fix beyond the scope of the PR)_.
- Copy-paste works with different quantization settings (as expected)

Hope this helps!

IanCaio commented:
Thanks for testing @russiankumar !

> - While using shortucts (eg. for muting - Ctrl+middle click) it still affects only the targeted TCO and not all the selected TCOs

True, I only changed the context menu actions, so the shortcuts were not affected. But I think this is an easy fix. I can change `mousePressEvent` to instead of calling `remove()` call `contextMenuAction( Remove )` for example. I personally think this is better than the other alternative I thought of, which would be moving the multiple/single remove logic to the `remove()` method, since I think that one should be reserved for individual removes. There's a third alternative which maybe would even make the `contextMenuAction` method much smaller, which is moving the logic from multiple/single remove/copy/cut/paste/mute to a new method which would check for selections and call the individual methods on each one. I would need to think of good naming for those though, maybe `removeSelected`, etc?

> - Right-click menu for paste works a bit weird. Say I want to paste somewhere but I hit the wrong box by mistake. In that case I would move my mouse and right click on the correct box. However I have to right click twice for this as the first right click only disables the previous menu. (On later testing, this is actually observed with every menu, so this is probably a fix beyond the scope of the PR).

Yeah, that's more related to how the Qt context menu works AFAIK, not sure the PR can really do anything to make it less annoying but if someone knows a way just let me know.

As for the pasting works, I understand it's not ideal because people probably expected to be able to paste anywhere on any tracks and everything would land on their appropriate tracks. However that's a deeper issue, because it involves how all other copy/paste operations works at LMMS at the moment: Currently, LMMS allows copying a track TCO and pasting it on another track, or dragging a whole selection of TCOs and dropping them on another track (as long as those rules mentioned before are followed). I even thought of making "Paste" just paste everything on its original track, but then I create a third behavior on the code base and contribute to the confusion. I thought following the guidelines of Drag&Drop was the ideal, making the individual TCO copy/paste the only operation that has a different behavior (which as mentioned could be fixed later by moving away from `Clipboard.cpp`). That's just my take on it, but I'd like to hear everyone's opinions as well!

IanCaio committed:
Separates methods for the actions on selections

Now the remove, copy, cut, paste and toggle mute actions have a separate method for applying them to selections, which are then called from the contextMenuAction method. That made the code more organized and the contextMenuAction method smaller.

Also, the mouse shortcuts for muting and removing (CTRL+middle button, middle button, CTRL+right button) now apply the action on selections as well.

Spekular commented on src/core/Clipboard.cpp:
```diff
@@ -35,6 +38,10 @@ void Clipboard::copy( JournallingObject * _obj )
+
+	// Clear the QApplication clipboard, so we don't have any conflicts when LMMS has to
+	// decide between the QApplication clipboard and the internal clipboard data
+	QApplication::clipboard()->clear( QClipboard::Clipboard );
```
This feels like an unnecessary/dangerous change. Have you checked all uses of the QApplication clipboard to ensure that this doesn't change copy/paste behavior at all?

IanCaio replied:
I didn't check every use of `QApplication::clipboard()`, but I'm not sure what's the danger in it. I mean, the clipboard is volatile, when you copy you add data to it but when you paste you can't trust the data will still be there (i.e.: if you copy something else before pasting). Maybe I'm missing something?

I had to do that because currently LMMS mixes the `Clipboard` class and the system's clipboard on copy/paste operations. The individual TCO copy uses `class Clipboard` but the multiple TCO copy uses the system's clipboard. So it's possible I'll have data on both simultaneously. When pasting I wouldn't be able to know which one should be pasted, so `Clipboard::copy` now clears the system's clipboard as well. Then I can know I should use it because the system's clipboard will be empty. Ideal would be if I cleared the `Clipboard` data as well when I copied multiple TCOs (which I don't currently do).

Another reason I advocate for changing everything to the system's clipboard, things can get confusing mixing both but I tried to come up with the cleanest solution possible, at least until the change happens (if everyone agrees). From a quick search it seems `Clipboard::copy` is only used on `src/core/Track.cpp` anyways, but other files uses `Clipboard::mimeType()` too.

Spekular commented on src/core/Track.cpp:
```diff
@@ -826,22 +830,50 @@ void TrackContentObjectView::mousePressEvent( QMouseEvent * me )
+			if( so.size() > 0 )
+			{
+				toggleMuteSelection( so );
+			}
```
I think a cleaner implementation than doing all these checks is to use one function for both selections and single clips. Simply pass a collection containing a single clip if you only want to affect that clip.

I'm also not sure that this logic holds. What if I select a bunch of clips, and the press mute on an unselected clip? `QVector so` will still contain the selected objects, so `so.size > 0` should be true, and the selection will be muted instead of the clip that was clicked on.

IanCaio replied:
Nice catch, you're right! I have to check both if there are selected TCO views and if the one where the event happened is among them. I'll fix that on all actions.

I thought about having `so.size > 0` inside the `toggleMuteSelection` method, but I thought the name wouldn't describe the behavior pretty well anymore, because `toggleMuteSelection` would both mute selections or individual tracks, and I don't know, it could be useful in the future to have those methods that take selections and modify them. What do you think about making a third method that would check if `so.size > 0` and if the event TCO view is selected them and then call either `toggleMuteSelection` or `toggleMute`? What could I call it?

IanCaio replied:
I fixed the bug and added the conditionals inside the `removeSelection()`,`copySelection()`,`cutSelection()`,`toggleMuteSelection()` methods. Still unsure about the naming, but had no better ideas until now :thinking:

IanCaio replied:
I fixed the bug and added the conditionals inside the `removeSelection()`,`copySelection()`,`cutSelection()` and `toggleMuteSelection()` methods. Still unsure about the naming, but had no better ideas until now :thinking:

Spekular commented:
I think the wording on the context menu items should change to reflect what action will be taken. Especially with the delete action it's good if users are made aware that the entire selection will be affected.

IanCaio commented:
> I think the wording on the context menu items should change to reflect what action will be taken. Especially with the delete action it's good if users are made aware that the entire selection will be affected.

Sure, could be done! Personally I think it's expected that "Delete" will remove everything selected. I understand some long time users could have gotten used to the current behavior though and since the `Undo` system is not flawless right now this could lead to headaches, so I'm open minded about changing it. Any ideas on what should be written instead? `Delete TCO/Selection` maybe?

IanCaio committed:
Fixes small bug and reorganize code

Fixes a small bug where using a mouse shortcut or choosing an action on a TCO that is not selected while other TCOs were selected would result in the selection being affected.

Also reorganizes the code so the conditional for choosing between the selection action and the individual action stays inside the method.

IanCaio committed:
Move logic to the action methods

Since the methods that called the action+Selection() methods didn't need the list of selectableObjects, I moved it to the inside of the action+Selection() methods and removed the parameter from them.

Spekular commented:
> Any ideas on what should be written instead?

I think "[action] selection" is great for selections. For single items I think we can probably leave the text as is. If you want to change it I'd prefer "[action] clip" over [action] TCO" as TCO is kind of a meaningless term to users (or should be, at least).

> I [...] added the conditionals inside the [action] methods.

Ah, that's not what I was proposing 😅. I meant more like this: 

```cpp
void TrackContentObjectView::mousePressEvent( QMouseEvent * me )
{
	//Get a list of selected selectableObjects
	QVector<selectableObject*> sos = gui->songEditor()->m_editor->selectedObjects();

	//Convert to a list of selected TCOVs
	QVector<TrackContentObjectView*> selection;
	selection.reserve(sos.size());
	for (auto so: sos) {
		TrackContentObjectView* tcov = dynamic_cast<TrackContentObjectView*>(*so);
		if (tcov != nullptr) { selection.append(tcov); }
	}

	//If we clicked part of the selection, affect all selected clips. Otherwise affect the clip we clicked
	QVector<TrackContentObjectView *> active = selection.contains(this) ? selection: QVector(this);

	//Now inside the switch we call
	removeSelection(active);
	togleMute(active);
	//etc...
}

The removeSelection, toggleMute and so on look a bit like this
...
void TrackContentObjectView::removeSelection(QVector<TrackContentObjectView *> tcovs)
{
	for (auto tcov: tcovs) { if (tcov != null) { tocv->remove(); } }
}
```

IanCaio committed:
Changes logic structure and labels

As suggested by Spekular, the logic structure was changed. Now, the mousePressEvent and contextMenuAction methods determine whether the action should happen to a selection of TCOs or an individual TCO. The list of TCOs to be affected populate a QVector list called "active", which is parsed to the action method that will apply the action to each object in the list.

I changed the method names to removeActive, cutActive, copyActive and toggleMuteActive, since I believe that better describe the behavior. The paste method is still called pasteSelection for now, because the paste behavior isn't related to the active TCOs but to the content of the clipboard.

The contextMenuEvent method on both src/core/Track.cpp and src/tracks/SampleTrack.cpp were changed so they also check if the right-clicked TCO is part of a selection or an individual TCO, and the labels for the actions are changed to better describe the behavior (i.e.: "Delete" for individual TCO and "Delete selection" for multiple TCOs).

IanCaio committed:
Make removeActive and toggleMuteActive static

removeActive and toggleMuteActive methods are now static so they can be called from anywhere in the code since they don't require a TCO view instance to work. That makes it possible for them to be used in the future if any feature requires this type of action to be called from out of a TCO view instance.

The same couldn't be done for the copyActive and cutActive methods because those require an instance of the TCO view to call them, since when copying to the clipboard some metadata is written using information from the object instance.

IanCaio commented:
@Spekular I updated the code to use the logic structure you suggested. I changed the name of the methods to `removeActive`, `copyActive`, `cutActive`, `toggleMuteActive`, which I think better describe the behavior now. The pasting method is still called `pasteSelection` (because that one is only called when we have a selection in the clipboard, otherwise `paste` is called).

The labels on the context menu were also updated so they change if the actions will affect selections. I also made `removeActive` and `toggleMuteActive` static as we discussed, so they can be called elsewhere in the code if future features require those actions to be called from outside a TCO view.

IanCaio commented on src/core/Track.cpp:
```diff
@@ -1125,25 +1165,170 @@ void TrackContentObjectView::contextMenuEvent( QContextMenuEvent * cme )
+			copyActive( active );
+			break;
+		case Paste:
+			// NOTE: Because we give preference to the QApplication clipboard over the LMMS Clipboard class, we need to
+			// clear the QApplication Clipboard during the LMMS Clipboard copy operations (Clipboard::copy does that)
+
+			// If we have TCO data on the clipboard paste it. If not, do our regular TCO paste.
+			if( QApplication::clipboard()->mimeData( QClipboard::Clipboard )->hasFormat( StringPairDrag::mimeType() ) )
+			{
+				pasteSelection();
+			}
+			else
+			{
+				getTrackContentObject()->paste();
+			}
```
What are your opinions on this logic staying here? Should I move it to the `pasteSelection` method and change its name to something like `pasteContent`? Or maybe keep it that way and just move everything to the `paste` method once we (possibly) move away from using `Clipboard::copy`?

IanCaio committed:
Renamed TCO View paste method

I renamed the TCO View paste method from pasteSelection to just paste, since the TCO view doesn't currently have a paste method (only the TCO class). It's also a name that accurately describes the action: it will paste either a group of TCOVs or a single TCOV on the current TCOV.

I also moved the logic for deciding between the multiple TCOV paste and single TCOV paste inside the paste method.

IanCaio commented:
Renamed `pasteSelection` to just `paste`, since I realized the TrackContentObjectView class doesn't have a paste method, just the TrackContentObject class does, hence it doesn't conflict with anything and is an appropriate name for the action.

Moved the logic for deciding between the multiple TCOV paste and individual TCOV paste to inside `TrackContentObjectView::paste()`

Spekular commented on src/core/Track.cpp:
```diff
@@ -735,8 +735,24 @@ void TrackContentObjectView::paintTextLabel(QString const & text, QPainter & pai
 {
-	// List of selected TCOs
-	QVector<selectableObject *> so = gui->songEditor()->m_editor->selectedObjects();
+	// Get a list of selected selectableObjects
+	QVector<selectableObject *> sos = gui->songEditor()->m_editor->selectedObjects();
+
+	// Convert to a list of selected TCOVs
+	QVector<TrackContentObjectView *> selection;
+	selection.reserve( sos.size() );
+	for( auto so: sos )
+	{
+		TrackContentObjectView *tcov = dynamic_cast<TrackContentObjectView *> ( so );
+		if( tcov != nullptr )
+		{
+			selection.append( tcov );
+		}
+	}
+
+	// If we clicked part of the selection, affect all selected clips. Otherwise affect the clip we clicked
+	QVector<TrackContentObjectView *> active = selection.contains( this ) ? selection : QVector<TrackContentObjectView *>( 1, this );
```
Let's extract this into a new method (`getClickedTCOs`?) since you reuse the same code later.

IanCaio replied:
Good idea! I updated the code to use this new method, which I'm calling `getClickedTCOs` as suggested for now

Spekular commented on src/core/Track.cpp:
```diff
@@ -1148,6 +1136,26 @@ void TrackContentObjectView::mouseReleaseEvent( QMouseEvent * me )
+	}
+
+	// Individual TCO or selection being right-clicked?
+	bool individualTCO = selection.contains( this ) ? false : true;
```
First of all, you don't need a ternary operator here, you can just use `!selection.contains(this)`.

Secondly, once the duplicate code is extracted, you can write this instead:
`bool individualTCO = getClickedTCOs().size <= 1;`
(Replacing `getClickedTCOs` with whatever we name the new method.)

Spekular commented on src/core/Track.cpp:
```diff
@@ -1179,177 +1197,134 @@ void TrackContentObjectView::contextMenuEvent( QContextMenuEvent * cme )
-	// List of selected TCOs
-	QVector<selectableObject *> so = gui->songEditor()->m_editor->selectedObjects();
+	// Get a list of selected selectableObjects
+	QVector<selectableObject *> sos = gui->songEditor()->m_editor->selectedObjects();
```
Another place where the newly extracted method should be used.

Spekular commented on src/tracks/SampleTrack.cpp:
```diff
@@ -343,6 +343,26 @@ void SampleTCOView::updateSample()
 
 void SampleTCOView::contextMenuEvent( QContextMenuEvent * _cme )
 {
+	// Depending on whether we right-clicked a selection or an individual TCO we will have
```
another extract

Spekular commented on include/Track.h:
```diff
@@ -250,12 +250,16 @@ class TrackContentObjectView : public selectableObject, public ModelView
-	void cutSelection( QVector<selectableObject *> so );
-	void pasteSelection(); // TODO: Rethink naming?
-	void toggleMuteSelection( QVector<selectableObject *> so );
+	// Methods to remove, copy, cut, paste and mute a QVector of TCO views
```
I think you should drop the `Active` from these names. Technically you can pass whatever QVector of TCOs you want and the function will copy/cut/remove/mute them, so I think it's more accurate that way.

IanCaio replied:
I also removed the `Active` from the method names, so they are now overloads of the individual TCOs methods. At first I thought they should be differentiated by name to make it less confusing to have the same name as other methods that works differently, but I agree the `QVector` argument should be enough for someone reading the code to understand what they are there for!

Spekular reviewed (changes requested):
Renaming the functions isn't strictly necessary (though I feel pretty strongly that the "Active" is unnecessary/inaccurate) but the repeated code definitely needs to be extracted into a new method.

IanCaio committed:
Merge branch 'master' into feature/selectedTCOActions

IanCaio committed:
Moves repeated code to a new method

This commit adds another method to TrackContentObjectView called getClickedTCOs, which will return a QVector with the TCO views that are supposed to be affected by a context menu action (either the individual right-clicked TCO or the selection of TCOs). Code was updated on the other methods to make use of this new method.

Method names for actions that affect multiple TCOs were changed. Instead of calling them copyActive, cutActive, toggleMuteActive and removeActive, they are just called copy, cut, toggleMute and remove, hence they are overloaded methods for those actions that affect multiple TCOs (their differenciation is the arguments list, which is a QVector list for those).

Veratil commented on src/core/Track.cpp:
```diff
@@ -734,6 +735,9 @@ void TrackContentObjectView::paintTextLabel(QString const & text, QPainter & pai
  */
 void TrackContentObjectView::mousePressEvent( QMouseEvent * me )
 {
+	QVector<TrackContentObjectView *> active = getClickedTCOs();
```
I wish there were a way to only use this in the `else if` blocks since this adds overhead to the left click event as well.

Spekular replied:
How about this?

`auto active = me->button() == Qt::RightButton ? getClickedTCOs(): QVector();`

IanCaio replied:
I'm not sure how this could be done without repeating code, because both `Qt::LeftButton` and `Qt::MidButton` use the `active` QVector at the moment.

It could be moved to inside the `if` blocks but we would repeat the `getClickedTCOs()` line 4 times. Or doing something similar to what @Spekular suggested and having `auto active = ( me->button() == Qt::LeftButton || Qt::MidButton ) ? QVector() : getClickedTCOs();`. Personally, between the two I think I prefer having `getClickedTCOs` repeated on each `if` block just before the action call, something like:

```
if( me->modifiers() & Qt::ControlModifier )
{
    active = getClickedTCOs();
    remove( active );
}
```

It doesn't look bad (even though the code is repeated a few times), would be more clear and avoids calling the `QVector()` constructor unnecessarily. Should I do it that way?

Spekular replied:
In case you saw the original comment and missed the edit, note that you don't need to check several modifiers. The options are left/mid/right, and active is used in mid and right, so checking for left is enough.

I also doubt the empty QVector() constructor takes much time at all.

If you *do* want to move the declaration into the if statements, please do it only once for right and once for middle click (i.e. two repetitions). There's no reason to do it once per action when all actions  in the branchrequire it.

IanCaio replied:
You're right, I misunderstood the logic when I commented. Your solution is drier and I'm up to using it! The idea of having the `getClickedTCOs` before the action was kind of getting ahead of myself thinking about if later we added actions that didn't use the `active` list but were activated by the Right/Mid buttons. Probably not worth it to get those extra lines thinking of something that might not even get added.

Veratil replied:
I'm cool with Spekular's way. 👍

Veratil commented on src/tracks/SampleTrack.cpp:
```diff
@@ -352,20 +356,30 @@ void SampleTCOView::contextMenuEvent( QContextMenuEvent * _cme )
+					tr( individualTCO
+						? "Delete (middle mousebutton)"
+						: "Delete selection (middle mousebutton)" ),
+						[this](){ contextMenuAction( Remove ); } );
```
This and all the other lambdas are indented as if they are part of the `tr()` function. This is confusing and should be de-dented.

Veratil commented on src/tracks/SampleTrack.cpp:
```diff
@@ -352,20 +356,30 @@ void SampleTCOView::contextMenuEvent( QContextMenuEvent * _cme )
 		contextMenu.addAction( embed::getIconPixmap( "cancel" ),
-					tr( "Delete (middle mousebutton)" ),
-						this, SLOT( remove() ) );
+					tr( individualTCO
```
Can we fix the indention of all the `addAction` calls? All the indention is crazy with unnecessary tabs.

Veratil commented on src/core/Track.cpp:
```diff
@@ -1125,25 +1133,176 @@ void TrackContentObjectView::contextMenuEvent( QContextMenuEvent * cme )
 		contextMenu.addAction( embed::getIconPixmap( "cancel" ),
-					tr( "Delete (middle mousebutton)" ),
-						this, SLOT( remove() ) );
+					tr( individualTCO
```
Can we fix the indention of all the addAction calls? All the indention is crazy with unnecessary tabs.

IanCaio replied:
Sure, I tried to keep the indentation as it was originally, but it can be fixed. Should it be this instead?

```
    contextMenu.addAction( embed::getIconPixmap( "cancel" ),
        tr( individualTCO ? "Delete (middle mousebutton) " : "Delete selection (middle mousebutton)" ),
        [this](){ contextMenuAction( remove ); } );
```

Veratil replied:
Second option is cleaner to me, just pull the `embed` part down as well.
```
contextMenu.addAction(
	embed,
	tr,
	[]()
);
```

Veratil commented on src/core/Track.cpp:
```diff
@@ -1125,25 +1133,176 @@ void TrackContentObjectView::contextMenuEvent( QContextMenuEvent * cme )
+					tr( individualTCO
+						? "Delete (middle mousebutton)"
+						: "Delete selection (middle mousebutton)" ),
+						[this](){ contextMenuAction( Remove ); } );
```
This and all the other lambdas are indented as if they are part of the tr() function. This is confusing and should be de-dented.

IanCaio committed:
Avoid unnecessary calls to getClickedTCOs()

We use a ternary operator inside TrackContentObjectView::mousePressEvent to avoid unnecessary calls to getClickedTCOs when the list is not going to be used.

The contextMenuEvent method on both Track.cpp and SampleTrack.cpp was reformated to use a more appropriate indentation and spacing between method calls.

IanCaio commented:
Fixed the indentation on `contextMenuEvent` methods and added blank lines for readability where convenient.

Avoided unnecessary calls to `getClickedTCOs` using @Spekular 's suggestion.

Veratil commented on src/core/Track.cpp:
```diff
@@ -735,8 +735,11 @@ void TrackContentObjectView::paintTextLabel(QString const & text, QPainter & pai
-	// active will be later used for the remove, copy, cut or toggleMute methods
+	// Right now, active is only used on right/mid clicks actions, so we use a ternary operator
+	// to avoid the overhead of calling getClickedTCOs when it's not used
+	auto active = me->button() == Qt::LeftButton
+							? QVector<TrackContentObjectView *>()
+							: getClickedTCOs();
```
I think two tabs over ternary options would be best, not 7 tabs.

Spekular replied:
It looks like you're using tabs for alignment, which isn't good. If someone displays tabs as 2 spaces, these lines wouldn't start to the right of the above. If they use 8 spaces, they would be *extremely* far out.

If you want to align the cases with the end of the row above, you'd want to tab until the start of the lines line up, and then use spaces from there. This preserves alignment between lines at all tab sizes.

However, I agree with Veratil that we could stick to just indenting one extra tab.

IanCaio replied:
My bad, even to the alignment I wanted I was off. I thought about aligning it with the conditional but I went further. I kept it as an extra tab as suggested.

IanCaio committed:
Fix indenting in a ternary operator assignment

Spekular commented:
@Veratil do you have any remaining complaints? If not, shall we merge this?

(Note: this will properly close #1212, IMO, so the relevant checkbox in the Better Workflow issue should be checked when it's merged)

Veratil commented:
> @Veratil do you have any remaining complaints? If not, shall we merge this?

LGTM 👍
